pxd_fastpath.c

```diff
@@ -0,0 +1,940 @@
+	atomic_inc(&pxd_dev->fp.ncount);
+	atomic_inc(&pxd_dev->fp.nswitch);
+
+	return 0;
+}
+
+static void _pxd_setup(struct pxd_device *pxd_dev, bool enable) {
+	if (!enable) {
+		printk(KERN_NOTICE "device %llu called to disable IO\n", pxd_dev->dev_id);
+		pxd_dev->connected = false;
+	} else {
+		printk(KERN_NOTICE "device %llu called to enable IO\n", pxd_dev->dev_id);
+	}
+
+	if (enable) {
+		spin_lock_irq(&pxd_dev->lock);
+		enableFastPath(pxd_dev, true);
+		spin_unlock_irq(&pxd_dev->lock);
+	}
+
+	if (enable) pxd_dev->connected = true;
+}
+
+void pxdctx_set_connected(struct pxd_context *ctx, bool enable) {
+	struct list_head *cur;
+	spin_lock(&ctx->lock);
+	list_for_each(cur, &ctx->list) {
+		struct pxd_device *pxd_dev = container_of(cur, struct pxd_device, node);
+
+		_pxd_setup(pxd_dev, enable);
+	}
+	spin_unlock(&ctx->lock);
+}
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
+static int __do_bio_filebacked(struct pxd_device *pxd_dev, struct bio *bio)
+{
+	loff_t pos;
+	unsigned int op = bio_op(bio);
+	int ret;
+	unsigned long startTime = jiffies;
+
+	// NOTE NOTE NOTE accessing out of lock
+	if (!pxd_dev->connected) {
+		printk(KERN_ERR"px is disconnected, failing IO.\n");
+		bio_io_error(bio);
+		return -EIO;
+	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,14,0)
+	generic_start_io_acct(pxd_dev->disk->queue, bio_op(bio), REQUEST_GET_SECTORS(bio), &pxd_dev->disk->part0);
+#else
+	generic_start_io_acct(bio_data_dir(bio), REQUEST_GET_SECTORS(bio), &pxd_dev->disk->part0);
+#endif
+
+	pxd_printk("do_bio_filebacked for new bio (pending %u)\n",
+				atomic_read(&pxd_dev->fp.ncount));
+	pos = ((loff_t) bio->bi_iter.bi_sector << 9);
+
+	switch (op) {
+	case REQ_OP_READ:
+		ret = pxd_receive(pxd_dev, bio, pos);
+		goto out;
+	case REQ_OP_WRITE:
+
+		if (bio->bi_opf & REQ_PREFLUSH) {
+			atomic_inc(&pxd_dev->fp.nio_preflush);
+			ret = _pxd_flush(pxd_dev);
+			if (ret < 0) goto out;
+		}
+
+		/* Before any newer writes happen, make sure previous write/sync complete */
+		pxd_check_write_cache_flush(pxd_dev);
+
+		ret = do_pxd_send(pxd_dev, bio, pos);
+		if (ret < 0) goto out;
+
+		if (bio->bi_opf & REQ_FUA) {
+			atomic_inc(&pxd_dev->fp.nio_fua);
+			ret = _pxd_flush(pxd_dev);
+			if (ret < 0) goto out;
+		}
+
+		ret = 0; goto out;
+
+	case REQ_OP_FLUSH:
+		ret = _pxd_flush(pxd_dev);
+		goto out;
+	case REQ_OP_DISCARD:
+	case REQ_OP_WRITE_ZEROES:
+		ret = _pxd_bio_discard(pxd_dev, bio, pos);
+		goto out;
+	default:
+		WARN_ON_ONCE(1);
+		ret = -EIO;
+		goto out;
+	}
+
+out:
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,14,0)
+	generic_end_io_acct(pxd_dev->disk->queue, bio_op(bio), &pxd_dev->disk->part0, startTime);
+#else
+	generic_end_io_acct(bio_data_dir(bio), &pxd_dev->disk->part0, startTime);
+#endif
+	atomic_inc(&pxd_dev->fp.ncomplete);
+	pxd_printk("Completed a request direction %p/%d\n", bio, bio_data_dir(bio));
+
+	if (ret < 0) {
+		bio_io_error(bio);
+		return ret;
+	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,3,0)
+	bio_endio(bio);
```

Comment:
There are too many places this is being done. I suggest wrapping in a macro an putting in pxd_compat.c